vendor/purs/runtime/pstring.ss

```diff
@@ -141,9 +276,12 @@
 
   ; Fast equality check based on object reference.
   (define (pstring-eq? x y)
-    (and (fx=? (pstring-length x) (pstring-length y))
-               (fx=? (pstring-offset x) (pstring-offset y))
-               (eq? (pstring-buffer x) (pstring-buffer y))))
+    (and
+      (Slice? x)
+      (Slice? y))
+      (and (fx=? (Slice-length x) (Slice-length y))
+                 (fx=? (Slice-offset x) (Slice-offset y))
+                 (eq? (Slice-buffer x) (Slice-buffer y))))
```

Comment:
I guess it would be annoying to compare ConcatTrees, but then this means that `concat`ting strings would prevent equality checks until they get compacted?

Reply:
Yeah, we compact the string if we can't do a reference check. It's quite difficult to say when would traversing the tree be faster than first compacting and then doing a (fast) walk over the slice. At least we could optimize the equality check so that it wouldn't compact if the tree is very short (like having only two chunks, prefix and suffix).

But now I notice you commented on `pstring-eq?` and not `pstring=?`. Maybe I missed your point?

Reply:
> But now I notice you commented on `pstring-eq?` and not `pstring=?`

It still applies - is a `pstring` always a `Slice`? The doc with the type sketch makes it sound like a `pstring` could also be a `Concat`. If that's the case then it would make sense to check for equality for the concat too

Reply:
Ah now I see what you mean. `pstring-eq?` is only for the reference check and that can only be used if you are comparing two slices. Notice that `pstring-eq?` is not exported.

Reply:
I renamed the function to avoid the confusion.

Reply:
Great, thanks!